README
CPAN-Testers-WWW-Statistics
===========================

Website creation tools for CPAN Tester Statistics.

Please note the tests for this code is very basic. If you wish to help me out
and can add to the test suite, please feel free.

INSTALLATION

Untar the tarball into the directory of choice. All scripts can then be run
from that directory. All the results end up in ./html, which can then be
copied to your web server.

DEPENDENCIES

The scripts require the following Perl modules to be installed:

  CGI
  DBI
  DBD::SQLite
  GD::Graph::lines
  Getopt::ArgvFile
  Getopt::Long
  HTML::Entities
  IO::File
  Template
  Time::Piece
  version
  WWW::Mechanize

SCRIPTS

* addresses.pl

Selects all the tester email addresses and attempts to guess the mapping of
any addresses not mapped to known people in the address.txt file.

* cpanstats-writepages

Web page creation script. Generates all the HTML files.

* cpanstats-writegraphs

Graph creation script. Generates all the graphs files.

* getmailrc.pl

Gets the latest copy of the 01mailrc.txt.gz and extracts the text file.

* updates.pl

Used in association with the cpanstats-data-latest.tar.gz tarball to update
templates and the RSS feed.

COPYRIGHT AND LICENCE

  Copyright (C) 2005-2008 Barbie for Miss Barbell Productions

  This package is free software; you can redistribute it and/or 
  modify it under the same terms as Perl itself.


Changes
Revision history for CPAN-Testers-WWW-Statistics
================================================

1.12    2014-06-02
        - added cpan/backpan 100 files to CPAN Stats page creation.
        - DateTime is now a pre-requisite.
        - added full version checks in META tests.
        - skip environment has not been configured.

1.11    2014-05-11
        - build performance data in isolation.

1.10    2014-04-21
        - tester() now returns an list.
        - monthstore and mainstore merged.
        - storage elements now saved in individual files.
        - dist lists written in lexical order.

1.09    2014-01-01
        - format of 01mailrc.txt has changed.

1.08    2014-01-01
        - fixed HTML rendering of names and email addresses.
        - dropped DBD::SQLite from prerequisites.

1.07    2013-12-27
        - added support for tester profiles.
        - added directory sizes (as suggested by David Golden).
        - reworked tester name lookups.

1.06    2013-08-18
        - template patches (thanks to Chorny).
        - set day and month names explicitly in Time::Piece.

1.05    2013-04-14
        - updated CSS.
        - fixed date tests.
        - provided additional logging.
        - extended the 'copy' files list.
        - parse title as HTML.

1.04    2013-03-31
        - add link to data file from html file for noreports.
        - RUNDATEs are based on DB timestamp not current timestamp (t/30setdates.t)
        - updated timestamps in templates to ease testing.
        - added minified CSS files.
        - added support for JSON loading of sponsors.

1.03    2013-01-13
        - abandoned support on Win32 until strftime in Time::Piece works.
        - updated license urls in META files.
        - added functionality to create CSV files for noreports feature.

1.02    2013-01-01
        - updated script:
          - bin/addresses.pl            (v0.17)
            - add metabase lookup.
            - switch to using phrasebook patterns.
            - added more legitimate 3/4 character gTLDs.
        - added new test script, 30setdates.t.
        - added new nreports functionality.
        - added sorttable.js - (c) Stuart Langridge.
        - better date testing.

1.01    2012-12-23
        - change file dates changed to meet W3CDTF standards.
        - provide skips if Google Chart API returns an error.
        - fixed labelling on percentage tables.
        - removed prototypes for alarm_handler.

1.00    2012-07-01
        - moved _tester_name() into parent as tester().
        - added Leaderboard management module.
        - updated POD
        - updated/new scripts:
          - bin/cpanstats-writegraphs   (v0.52)
          - bin/cpanstats-writepages    (v0.52)
          - bin/cpanstats-leaderboard   (v0.01)

0.99    2012-04-22
        - META tests no longer require specific versions.
        - fixed pre-requisite lists.

0.98    2012-04-22
        - SEE ALSO package name fixes.
        - reworded index and failure pages.
        - removed requirement on local SQLite database.
        - spelling fixes. (thanks to Florian Schlichting & Gregor Herrmann).
        - changes to scripts with above fixes:
          - bin/addresses.pl            (v0.16)
          - bin/cpanstats-writegraphs   (v0.51)
          - bin/cpanstats-writepages    (v0.51)
          - bin/getmailrc.pl            (v0.03)
        - removed files, as no longer required:
          - t/data/21config08.ini
          - t/data/21config09.ini
        - added CPAN Testers Sponsors and Perl adverts.
        - added minimum perl version (5.006).
        - fixed test pre-requisites.
        - reworked Makefile.PL for clarity.

0.97    2012-01-31
        - final leaderboard fixes (os and total).

0.96    2012-01-30
        - set_max & set_range reworked (Graph.pm).
        - fix dates for test database creation.
        - fixed range limits.
        - remove social media bar.

0.95    2012-01-26
        - removed NNTP code from cpanmail.cgi.
        - added cpanmail.ini to cgi templates.
        - don't overwrite basic files in case they have changed.
        - added the ability to link basic files.
        - changed date names
        - leaderboard fixes.

0.94    2011-08-21
        - revised banner logo image by Jon 'JJ' Allen.
        - update to MANIFEST to include missing layout images.
        - added addresses.pl example shell scripts.
        - added cpanstats-admin.pl to examples.
        - removed obsolete updates.pl script.
        - redesigned _build_monthly_stats() to use JSON file storage.
        - fixed tests for Win32 (patches by Christian Walde).
        - fixed main leaderboard listings.
        - fixed first/last testers stats.

0.93    2011-07-13
        - added more filters for 'no reports' in example config.
        - update layout to include an update notification.
        - added new social media bar

0.92    2011-07-09
        - added filters for distros that should not appear in the 'no reports' 
          list.

0.91    2011-07-04
        - updated favicon links.
        - added 'no report' functionality.
        - made path to 01mailrc.txt a configurable parameter.

0.90    2011-07-03
        - fixes to prevent blank oses being displayed in leaderboards.
        - adjustments to postdates in leaderboards.

0.89    2011-07-03

Makefile.PL
use ExtUtils::MakeMaker;

WriteMakefile(
    'NAME'             => 'CPAN::Testers::WWW::Statistics',
    'VERSION_FROM'     => 'lib/CPAN/Testers/WWW/Statistics.pm',
    'PREREQ_PM'        => {

        DBI                     => 0,
        DBD::SQLite             => 1.07,
        GD::Graph::lines        => 0,
        File::Basename          => 0,
        File::Copy              => 0,
        File::Path              => 0,
        Getopt::ArgvFile        => 0,
        Getopt::Long            => 0,
        HTML::Entities          => 0,
        IO::File                => 0,
        Sort::Versions          => 0,
        Template                => 0,
        WWW::Mechanize          => 0,

        # for the CGI tools
        CGI                     => 0,
        Time::Piece             => 0,
    },

    NO_META            => 1,

    ($] >= 5.005 ?    ## Add these new keywords supported since 5.005
      (ABSTRACT => 'Scripts to create the cpanstats website',
       AUTHOR     => 'Barbie <barbie@cpan.org>') : ()),
);


lib/CPAN/Testers/WWW/Statistics.pm
package CPAN::Testers::WWW::Statistics;

use warnings;
use strict;
use vars qw($VERSION);

$VERSION = '0.49';

#----------------------------------------------------------------------------

=head1 NAME

CPAN::Testers::WWW::Statistics - CPAN Testers Statistics website.

=head1 DESCRIPTION

CPAN Testers Statistics comprises the actual website pages, a CGI tool to find
testers, and some back end code to help map tester address to a real identity.

=cut

q("I am NOT a number!");

__END__

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties, that is not explained within the POD
documentation, please send bug reports and patches to the RT Queue (see below).

Fixes are dependant upon their severity and my availablity. Should a fix not
be forthcoming, please feel free to (politely) remind me.

RT Queue -
http://rt.cpan.org/Public/Dist/Display.html?Name=CPAN-Testers-WWW-Statistics

=head1 SEE ALSO

L<CPAN::WWW::Testers::Generator>,
L<CPAN::WWW::Testers>

F<http://www.cpantesters.org/>,
F<http://stats.cpantesters.org/>,
F<http://wiki.cpantesters.org/>

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT AND LICENSE

  Copyright (C) 2008 Barbie for Miss Barbell Productions.

  This module is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.

=cut



lib/CPAN/Testers/WWW/Statistics/Database.pm
package CPAN::Testers::WWW::Statistics::Database;

use warnings;
use strict;
use vars qw($VERSION);

$VERSION = '0.49';

#----------------------------------------------------------------------------

=head1 NAME

CPAN::Testers::WWW::Statistics::Database - DB handling code.

=head1 SYNOPSIS

  my $dbi = CPAN::Testers::WWW::Statistics::Database->new(database => $db);
  my @rows = $dbi->get_query($sql);
  $dbi->do_query($sql);

  my $iterator = $dbi->get_query_interator($sql);
  while(my $row = $iterator->()) {
    # do something
  }

=head1 DESCRIPTION

Database handling code for interacting with a local cpanstats database.

=cut

# -------------------------------------
# Library Modules

use DBI;

# -------------------------------------
# Variables

use constant    DATABASE    => 'cpanstats.db';

# -------------------------------------
# Routines

=head1 INTERFACE

=head2 The Constructor

=over 4

=item * new

=back

=cut

sub new {
    my ($class,%hash) = @_;
    my $self = {};

    $self->{database} = $hash{database} || DATABASE;

    $self->{dbh} = DBI->connect("dbi:SQLite:dbname=$self->{database}",'','');
    $self->{dbh}->{AutoCommit} = 1;

    bless $self, $class;
    return $self;
}

sub DESTROY {
    my $self = shift;
    return      unless($self->{dbh});

    $self->{dbh}->disconnect;
}

=head2 Methods

=over 4

=item * do_query

An SQL wrapper method to perform a non-returning request.

=cut

sub do_query {
    my ($self,$sql) = @_;

    # prepare the sql statement for executing
    my $sth = $self->{dbh}->prepare($sql);
    print STDERR $self->{dbh}->errstr.":$sql\n"	unless($sth);

    # execute the SQL using any values sent to the function
    # to be placed in the sql
    if(!$sth->execute()) {
        print STDERR $sth->errstr,$sql
    }

    $sth->finish;
}

=item * get_query

An SQL wrapper method to perform a returning request.

=cut

sub get_query {
    my ($self,$sql) = @_;
    my @rows;
    my $sth = $self->{dbh}->prepare($sql);
    $sth->execute;
    while(my $row = $sth->fetchrow_arrayref) {
        push @rows, [@$row];
    }
    return @rows;
}

=item * get_query_iterator

An SQL wrapper method to perform a returning request, via an iterator.

=cut

sub get_query_iterator {
    my ($self,$sql) = @_;
    my @rows;
    my $sth = $self->{dbh}->prepare($sql);
    $sth->execute;

    return sub { return $sth->fetchrow_arrayref }
}

q("I code for beer!");

__END__

=back

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties, that is not explained within the POD
documentation, please send bug reports and patches to the RT Queue (see below).

Fixes are dependant upon their severity and my availablity. Should a fix not
be forthcoming, please feel free to (politely) remind me.

RT Queue -
http://rt.cpan.org/Public/Dist/Display.html?Name=CPAN-Testers-WWW-Statistics

=head1 SEE ALSO

L<CPAN::WWW::Testers::Generator>,
L<CPAN::WWW::Testers>

F<http://www.cpantesters.org/>,
F<http://stats.cpantesters.org/>,
F<http://wiki.cpantesters.org/>

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT AND LICENSE

  Copyright (C) 2008 Barbie for Miss Barbell Productions.

  This module is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.

=cut



lib/CPAN/Testers/WWW/Statistics/Pages.pm
package CPAN::Testers::WWW::Statistics::Pages;

use warnings;
use strict;
use vars qw($VERSION);

$VERSION = '0.49';

#----------------------------------------------------------------------------

=head1 NAME

CPAN::Testers::WWW::Statistics::Pages - CPAN Testers Statistics pages.

=head1 SYNOPSIS

  my $ct = CPAN::Testers::WWW::Statistics::Pages->new(progress => 1);
  $ct->create();

=head1 DESCRIPTION

Using the cpanstats database, this module extracts all the data and generates
all the HTML pages needed for the CPAN Testers Statistics website. In addition,
also generates the data files in order generate the graphs that appear on the
site.

=cut

# -------------------------------------
# Library Modules

use File::Basename;
use File::Copy;
use File::Path;
use HTML::Entities;
use IO::File;
use Sort::Versions;
use Template;
#use Time::HiRes qw ( time );

use CPAN::Testers::WWW::Statistics::Database;

# -------------------------------------
# Variables

use constant DATABASE => './cpanstats.db';
use constant ADDRESS  => 'data/addresses.txt';

my ($known_s,$known_t) = (0,0);

my %month = (
    0 => 'January',   1 => 'February', 2 => 'March',     3 => 'April',
    4 => 'May',       5 => 'June',     6 => 'July',      7 => 'August',
    8 => 'September', 9 => 'October', 10 => 'November', 11 => 'December'
);

my ($LIMIT,@versions,%options,%pages);
my ($THISYEAR,$RUNDATE,$STATDATE,$THISDATE,$THATYEAR,$LASTDATE,$THATDATE);
my ($DATABASE2);

# -------------------------------------
# Subroutines

=head1 INTERFACE

=head2 The Constructor

=over 4

=item * new

Page creation object. Allows the user to turn or off the progress tracking.

new() takes an option hash as an argument, which may contain 'progress => 1'
to turn on the progress tracker and/or 'database => $db' to indicate the path
to the database. If no database path is supplied, './cpanstats.db' is used.

=back

=cut

sub new {
    my ($class,%hash) = @_;
    my $self = {};

    $self->{templates} = $hash{templates} || './templates';
    $self->{directory} = $hash{directory} || './html';
    $self->{progress}  = $hash{progress}  || 0;
    $self->{database}  = $hash{database}  || DATABASE;
    $self->{address}   = $hash{address}   || ADDRESS;

    bless $self, $class;

    return unless(-f $self->{database});
    return $self;
}

=head2 Public Methods

=over 4

=item * create

Method to facilitate the creation of graphs.

=cut

sub create {
    my $self = shift;

    _progress("init") if($self->{progress});
    _init_date();

    $self->{dbh} = CPAN::Testers::WWW::Statistics::Database->new(database => $self->{database});

    _progress("start") if($self->{progress});
    $self->_write_basics();
    $self->_write_stats();
    _progress("finish") if($self->{progress});
}

=back

=head2 Private Methods

=over 4

=item * _write_basics

Write out basic pages, all of which are simply built from the templates,
without any data processing required.

=cut

sub _write_basics {
    my $self = shift;

    mkpath($self->{directory});

    # calculate database metrics
    my $mtime = (stat($self->{database}))[9];
    my @ltime = localtime($mtime);
    $DATABASE2 = sprintf "%d%s %s %d", $ltime[3],_ext($ltime[3]),$month{$ltime[4]},$ltime[5]+1900;
    my $DATABASE1 = sprintf "%04d/%02d/%02d", $ltime[5]+1900,$ltime[4]+1,$ltime[3];
    my $DBSZ_UNCOMPRESSED = int((-s $self->{database}        ) / (1024 * 1024));
    my $DBSZ_COMPRESSED   = int((-s $self->{database} . '.gz') / (1024 * 1024));

    # additional pages not requiring metrics
    my %pages = (
            index    => {THISDATE => $THISDATE, DATABASE => $DATABASE1,
                         DBSZ_COMPRESSED => $DBSZ_COMPRESSED, DBSZ_UNCOMPRESSED => $DBSZ_UNCOMPRESSED},
            updates  => {},
            cpanmail => {},
            response => {},
            graphs   => {});

    _progress("building support pages") if($self->{progress});
    $self->_writepage($_,$pages{$_})    for(keys %pages);

    # copy files
    foreach my $filename (
        'rss-2.0.xml',
        'cgi-bin/cpanmail.cgi',
        'favicon.ico',
        'style.css',
        map {'images/'.basename($_)} glob($self->{templates} . '/images/*'),
        )
    {
        my $src  = "$self->{templates}/$filename";
        my $dest = "$self->{directory}/$filename";
        mkpath( dirname($dest) );
        copy( $src, $dest );
    }
}

=item * _write_stats

Extracts data, compiles the pages, generates the graph data files and
creates the HTML pages.

=cut

sub _write_stats {
    my $self = shift;
    my (%stats,%perls,%fails,%pass,%tvars);

    _progress("building report matrix") if($self->{progress});

    $self->_report_matrix();

## BUILD GENERAL STATS

    _progress("building stats hash") if($self->{progress});

    my $iterator = $self->{dbh}->get_query_iterator("SELECT * FROM cpanstats");
    while(my $row = $iterator->()) {
        #insert_report($id,$state,$date,$from,$distvers,$platform);
        $row->[7] =~ s/\s.*//;  # only need to know the main release

        if($row->[1] eq 'cpan') {
            $stats{$row->[2]}->{pause}++;
            $stats{$row->[2]}->{uploads} ->{$row->[4]}
                                         ->{$row->[5]}++;
            $fails{$row->[4]}->{$row->[5]}->{post} = $row->[2];
        } else {
            my $name = $self->_tester_name($row->[3]);
            $stats{$row->[2]}->{reports}++;
            $stats{$row->[2]}->{state}   ->{$row->[1]}++;
            $stats{$row->[2]}->{tester}  ->{$name    }++;
            $stats{$row->[2]}->{dist}    ->{$row->[4]}++;
            $stats{$row->[2]}->{version} ->{$row->[5]}++;
            $stats{$row->[2]}->{platform}->{$row->[6]}++;
            $stats{$row->[2]}->{perl}    ->{$row->[7]}++;

            # check failure rates
            $fails{$row->[4]}->{$row->[5]}->{fail}++    if($row->[1] =~ /FAIL|UNKNOWN/i);
            $fails{$row->[4]}->{$row->[5]}->{pass}++    if($row->[1] =~ /PASS/i);
            $fails{$row->[4]}->{$row->[5]}->{total}++;

            # build latest pass matrix
            my $perl = $row->[7];
            $perl =~ s/\s.*//;  # only need to know the main release
            $perls{$perl} = 1;
            $pass{$row->[6]}->{$perl}->{$row->[4]} = 1;
        }
    }

    _progress("stats hash built") if($self->{progress});

## GENERATE DISTRIBUTION MATRIX

    _progress("building distribution matrix") if($self->{progress});

    my $index = 0;
    my $content = '<table id="matrix">';
    $content .= '<tr><th>Platform/Perl</th><th>' . join("</th><th>",@versions) . '</th></tr>';
    for my $platform (sort keys %pass) {
        $content .= '<tr><th>' . $platform . '</th>';
        for my $perl (@versions) {
            my $count = scalar(keys %{$pass{$platform}->{$perl}}) || 0;
            if($count) {
                $index++;

                my @list = sort keys %{$pass{$platform}->{$perl}};
                $tvars{dists}     = \@list;
                $tvars{vplatform} = $platform;
                $tvars{vperl}     = $perl;
                $tvars{count}     = $count;
                $self->_writepage('list-'.$index,\%tvars);
                undef %tvars;
            }

            $content .= '<td class='
                        . ($count > 50 ? 'lots' :
                          ($count >  0 ? 'some' : 'none'))
                        . '>'
                        . ($count ? qq|<a href="list-$index.html">$count</a>| : '-')
                        . '</td>';
        }
        $content .= '</tr>';
    }
    $content .= '</table>';

    $tvars{CONTENT} = $content;
    $self->_writepage('dmatrix',\%tvars);
    undef %tvars;

## GENERATE MONTHLY REPORT STAT GRAPHS & TABLE

    _progress("building monthly stats for graphs") if($self->{progress});

    #print "DATE,UPLOADS,REPORTS,NA,PASS,FAIL,UNKNOWN\n";
    my $fh = IO::File->new(">stats1.txt");
    my $fh3 = IO::File->new(">stats3.txt");
    for my $date (sort keys %stats) {
        next    if($date > $LIMIT);
        my @fields = (
            $date,
            $stats{$date}->{pause},
            $stats{$date}->{reports},
            $stats{$date}->{state}->{pass},
            $stats{$date}->{state}->{fail}
        );

        unshift @{$tvars{STATS}},
            [   @fields,
                $stats{$date}->{state}->{na},
                $stats{$date}->{state}->{unknown}];

        # graphs don't include current month
        next    if($date > $LIMIT-1);

        my $content = sprintf "%d,%d,%d,%d,%d\n", @fields;
        print $fh $content;

        $content = sprintf "%d,%d,%d,%d\n",
            $date,
            $stats{$date}->{state}->{fail},
            $stats{$date}->{state}->{na},
            $stats{$date}->{state}->{unknown};
        print $fh3 $content;
    }
    $fh->close;
    $fh3->close;

    $self->_writepage('mreports',\%tvars);
    undef %tvars;

## GENERATE MONTHLY REPORT STAT GRAPHS

    #print "DATE,TESTERS,PLATFORMS,PERLS\n";
    $fh = IO::File->new(">stats2.txt");
    for my $date (sort keys %stats) {
        next    if($date > $LIMIT-1);
        printf $fh "%d,%d,%d,%d\n",
            $date,
            scalar(keys %{$stats{$date}->{tester}}),
            scalar(keys %{$stats{$date}->{platform}}),
            scalar(keys %{$stats{$date}->{perl}});
    }
    $fh->close;

## GENERATE FAILURE RATE TABLES

    _progress("building failure rates") if($self->{progress});

    # calculate worst failure rates - by failure count
    my %worst;
    for my $dist (keys %fails) {
        my ($version) = sort {versioncmp($b,$a)} keys %{$fails{$dist}};
        $worst{"$dist-$version"} = $fails{$dist}->{$version};
        $worst{"$dist-$version"}->{dist}   = $dist;
        $worst{"$dist-$version"}->{pcent}  = $fails{$dist}->{$version}->{fail} ? int(($fails{$dist}->{$version}->{fail}/$fails{$dist}->{$version}->{total})*10000)/100 : 0.00;
        $worst{"$dist-$version"}->{pass} ||= 0;
    }
    my $count = 1;
    for my $dist (sort {$worst{$b}->{fail} <=> $worst{$a}->{fail} || $worst{$b}->{pcent} <=> $worst{$a}->{pcent}} keys %worst) {
        last unless($worst{$dist}->{fail});
        my $pcent = sprintf "%3.2f%%", $worst{$dist}->{pcent};
        push @{$tvars{WORST}}, [$count++, $worst{$dist}->{fail}, $dist, $worst{$dist}->{post}, $worst{$dist}->{pass}, $worst{$dist}->{total}, $pcent, $worst{$dist}->{dist}];
        last    if($count > 100);
    }

    $tvars{DATABASE} = $DATABASE2;
    $self->_writepage('wdists',\%tvars);
    undef %tvars;

    # calculate worst failure rates - by percentage
    $count = 1;
    for my $dist (sort {$worst{$b}->{pcent} <=> $worst{$a}->{pcent} || $worst{$b}->{fail} <=> $worst{$a}->{fail}} keys %worst) {
        last unless($worst{$dist}->{fail});
        my $pcent = sprintf "%3.2f%%", $worst{$dist}->{pcent};
        push @{$tvars{WORST}}, [$count++, $worst{$dist}->{fail}, $dist, $worst{$dist}->{post}, $worst{$dist}->{pass}, $worst{$dist}->{total}, $pcent, $worst{$dist}->{dist}];
        last    if($count > 100);
    }

    $tvars{DATABASE} = $DATABASE2;
    $self->_writepage('wpcent',\%tvars);
    undef %tvars;

## GENERATE MONTHLY TESTERS / PLATFORM / PERL REPORT TABLES

    _progress("building monthly tables") if($self->{progress});

    for my $date (sort keys %stats) {
        next    if($date > $LIMIT);

        my ($count,$content) = (0,'');
        for my $platform (sort {$stats{$date}->{platform}->{$b} <=> $stats{$date}->{platform}->{$a}} keys %{$stats{$date}->{platform}}) {
            $content .= ', '    if($content);
            $content .= "[$stats{$date}->{platform}->{$platform}] $platform";
            $count++;
        }
        unshift @{$tvars{STATS}}, [$date,$count,$content];
    }
    $self->_writepage('mplatforms',\%tvars);
    undef %tvars;

    for my $date (sort keys %stats) {
        next    if($date > $LIMIT);

        my ($count,$content) = (0,'');
        for my $perl (sort {$stats{$date}->{perl}->{$b} <=> $stats{$date}->{perl}->{$a}} keys %{$stats{$date}->{perl}}) {
            $content .= ', '    if($content);
            $content .= "[$stats{$date}->{perl}->{$perl}] $perl";
            $count++;
        }
        unshift @{$tvars{STATS}}, [$date,$count,$content];
    }
    $self->_writepage('mperls',\%tvars);
    undef %tvars;

    my %testers;
    for my $date (sort keys %stats) {
        next    if($date > $LIMIT);

        my ($count,$content) = (0,'');
        for my $tester (sort {$stats{$date}->{tester}->{$b} <=> $stats{$date}->{tester}->{$a}} keys %{$stats{$date}->{tester}}) {
            $content .= ', '    if($content);
            $content .= "[$stats{$date}->{tester}->{$tester}] $tester";
            $testers{$tester} += $stats{$date}->{tester}->{$tester};
            $count++;
        }
        unshift @{$tvars{STATS}}, [$date,$count,$content];
    }
    $self->_writepage('mtesters',\%tvars);
    undef %tvars;

## GENERATE TESTERS LEADER BOARD

    _progress("building leader board") if($self->{progress});

    $count = 1;
    for my $tester (sort {$testers{$b} <=> $testers{$a}} keys %testers) {
        push @{$tvars{STATS}}, [$count++, $testers{$tester}, $tester];
    }

    $count--;
    print "Unknown Addresses: ".($count-$known_t)."\n";
    print "Known Addresses:   ".($known_s)."\n";
    print "Listed Addresses:  ".($known_s+$count-$known_t)."\n";
    print "\n";
    print "Unknown Testers:   ".($count-$known_t)."\n";
    print "Known Testers:     ".($known_t)."\n";
    print "Listed Testers:    ".($count)."\n";

    push @{$tvars{COUNTS}}, ($count-$known_t),$known_s,($known_s+$count-$known_t),($count-$known_t),$known_t,$count;

    $self->_writepage('testers',\%tvars);
    undef %tvars;
    undef %testers;
}

=item * _report_matrix

Generates the report matrix page

=cut

sub _report_matrix {
    my $self = shift;
    my (%stats,%perls,%tvars);

    # grab all records for the month
    my $iterator = $self->{dbh}->get_query_iterator("SELECT * FROM cpanstats WHERE postdate like '$LASTDATE'");
    while(my $row = $iterator->()) {
        next    if($row->[1] eq 'cpan');

        # platform -> perl
        $row->[7] =~ s/\s.*//;  # only need to know the main release
        $stats{$row->[6]}->{$row->[7]}++;
        $perls{$row->[7]} = 1;
    }

    my @versions = sort {versioncmp($b,$a)} keys %perls;

    my $content = '<table id="matrix">';
    $content .= '<tr><th>Platform/Perl</th><th>' . join("</th><th>",@versions) . '</th></tr>';
    for my $platform (sort keys %stats) {
        $content .= '<tr><th>' . $platform . '</th>';
        for my $perl (@versions) {
            $content .= '<td class='
                        . ($stats{$platform}->{$perl} > 50 ? 'lots' :
                          ($stats{$platform}->{$perl} >  0 ? 'some' : 'none'))
                        . '>'
                        . ($stats{$platform}->{$perl} || '-')
                        . '</td>';
        }
        $content .= '</tr>';
    }
    $content .= '</table>';

    $tvars{CONTENT} = $content;
    $self->_writepage('rmatrix',\%tvars);
}

=item * _writepage

Creates a single HTML page.

=cut

sub _writepage {
    my ($self,$page,$vars) = @_;

#   _progress("writing page - $page") if($self->{progress});

    my $template = $page;
    $template = 'distlist'    if($page =~ /^list-/);

    my $layout = "layout.html";
    my $source = "$template.html";
    my $target = "$self->{directory}/$page.html";

    $vars->{SOURCE}   = $source;
    $vars->{VERSION}  = $VERSION;
    $vars->{RUNDATE}  = $RUNDATE;
    $vars->{STATDATE} = $STATDATE;
    $vars->{THATDATE} = $THATDATE;

#    if($page =~ /^list-/) {
#        use Data::Dumper;
#        print STDERR "$page:" . Dumper($pages{$page});
#    }

    my %config = (                              # provide config info
        RELATIVE        => 1,
        ABSOLUTE        => 1,
        INCLUDE_PATH    => $self->{templates},
        INTERPOLATE     => 0,
        POST_CHOMP      => 1,
        TRIM            => 1,
    );

    my $parser = Template->new(\%config);           # initialise parser
    $parser->process($layout,$vars,$target) # parse the template
        or die $parser->error();
}

=item * _tester_name

Returns either the known name of the tester for the given email address, or
returns a doctored version of the address for displaying in HTML.

=cut

my $address;
sub _tester_name {
    my ($self,$name) = @_;

    $address ||= do {
        my (%address_map,%known);
        my $fh = IO::File->new($self->{address})    or die "Cannot open address file [$self->{address}]: $!";
        while(<$fh>) {
            chomp;
            my ($source,$target) = (/(.*),(.*)/);
            next    unless($source && $target);
            $address_map{$source} = $target;
            $known{$target}++;
        }
        $fh->close;
        $known_t = scalar(keys %known);
        $known_s = scalar(keys %address_map);
        \%address_map;
    };

    my $addr = ($address->{$name} && $address->{$name} =~ /\&\#x?\d+\;/)
                ? $address->{$name}
                : encode_entities( ($address->{$name} || $name) );
    $addr =~ s/\./ /g if($addr =~ /\@/);
    $addr =~ s/\@/[]/g;
    $addr =~ s/</&lt;/g;
    return $addr;
}

=item * _ext

Provides the ordinal for dates.

=cut

sub _ext {
    my $num = shift;
    return 'st' if($num == 1 || $num == 21 || $num == 31);
    return 'nd' if($num == 2 || $num == 22);
    return 'rd' if($num == 3 || $num == 23);
    return 'th';
}

=item * _progress

Simple audit logging function.

=cut

my $lasttime = time;

sub _progress {
    my $msg = shift;
    my $time = time;
    my @localtime = localtime($time);
    my $secs = $time - $lasttime;
    printf STDERR "%02d:%02d:%02d\t%03d\t%s\n", $localtime[2], $localtime[1], $localtime[0], $secs, $msg;
    $lasttime = $time;
}

=item * _init_date

Prime all key date variable.

=cut

sub _init_date {
    my @datetime = localtime;
    $THISYEAR = ($datetime[5] +1900);
    $RUNDATE  = sprintf "%d%s %s %d",
                        $datetime[3], _ext($datetime[3]),
                        $month{$datetime[4]}, $THISYEAR;

    # LIMIT is the last date for all data
    $LIMIT    = ($THISYEAR) * 100 + $datetime[4] + 1;
    if($datetime[4] == 0) {
        $datetime[4] = 12;
        $THISYEAR--;
    }

    # STATDATE/THISDATE is the Month/Year stats are run for
    $STATDATE = sprintf "%s %d", $month{int($datetime[4])}, $THISYEAR;
    $THISDATE = sprintf "%04d%02d", $THISYEAR, int($datetime[4]);

    # LASTDATE/THATDATE is the previous Month/Year for a full matrix
    $datetime[4]--;
    $THATYEAR = $THISYEAR;
    if($datetime[4] == 0) {
        $datetime[4] = 11;
        $THATYEAR--;
    }
    $LASTDATE = sprintf "%04d%02d", $THATYEAR, int($datetime[4]);
    $THATDATE = sprintf "%s %d", $month{int($datetime[4])}, $THISYEAR;

    #print STDERR "THISYEAR=[$THISYEAR]\n";
    #print STDERR "LIMIT=[$LIMIT]\n";
    #print STDERR "STATDATE=[$STATDATE]\n";
    #print STDERR "RUNDATE=[$RUNDATE]\n";
}

q("I'm breaking the back of love.");

__END__

=back

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties, that is not explained within the POD
documentation, please send bug reports and patches to the RT Queue (see below).

Fixes are dependant upon their severity and my availablity. Should a fix not
be forthcoming, please feel free to (politely) remind me.

RT Queue -
http://rt.cpan.org/Public/Dist/Display.html?Name=CPAN-Testers-WWW-Statistics

=head1 SEE ALSO

L<CPAN::WWW::Testers::Generator>,
L<CPAN::WWW::Testers>

F<http://www.cpantesters.org/>,
F<http://stats.cpantesters.org/>,
F<http://wiki.cpantesters.org/>

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT AND LICENSE

  Copyright (C) 2005-2008 Barbie for Miss Barbell Productions.

  This module is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.

=cut

